hexkit/protocols/objstorage.py

```diff
@@ -0,0 +1,639 @@
+        """Thrown when trying to access a bucket with an ID that doesn't exist."""
+
+        def __init__(self, bucket_id: Optional[str]):
+            message = (
+                "The bucket "
+                + (f"with ID '{bucket_id}' " if bucket_id else "")
+                + "does not exist."
+            )
+            super().__init__(message)
+
+    class BucketAlreadyExistsError(BucketError):
+        """Thrown when trying to create a bucket with an ID that already exists."""
+
+        def __init__(self, bucket_id: Optional[str]):
+            message = (
+                "The bucket "
+                + (f"with ID '{bucket_id}' " if bucket_id else "")
+                + "already exist."
+            )
+            super().__init__(message)
+
+    class ObjectError(ObjectStorageProtocolError):
+        """Generic base exceptions for error that occur while handling file objects."""
+
+    class ObjectNotFoundError(ObjectError):
+        """Thrown when trying to access a bucket with an ID that doesn't exist."""
+
+        def __init__(
+            self, bucket_id: Optional[str] = None, object_id: Optional[str] = None
+        ):
+            message = (
+                "The object "
+                + (f"with ID '{object_id}' " if object_id else "")
+                + (f"in bucket with ID '{bucket_id}' " if bucket_id else "")
+                + "does not exist."
+            )
+            super().__init__(message)
+
+    class ObjectAlreadyExistsError(ObjectError):
+        """Thrown when trying to access a file with an ID that doesn't exist."""
+
+        def __init__(
+            self, bucket_id: Optional[str] = None, object_id: Optional[str] = None
+        ):
+            message = (
+                "The object "
+                + (f"with ID '{object_id}' " if object_id else "")
+                + (f"in bucket with ID '{bucket_id}' " if bucket_id else "")
+                + "already exist."
+            )
+            super().__init__(message)
+
+    class BucketIdValidationError(BucketError):
+        """Thrown when a bucket ID is not valid."""
+
+        def __init__(self, bucket_id: str, reason: Optional[str]):
+            message = f"The specified bucket ID '{bucket_id}' is not valid" + (
+                f": {reason}." if reason else "."
+            )
+            super().__init__(message)
+
+    class ObjectIdValidationError(ObjectError):
+        """Thrown when an object ID is not valid."""
+
+        def __init__(self, object_id: str, reason: Optional[str]):
+            message = f"The specified object ID '{object_id}' is not valid" + (
+                f": {reason}." if reason else "."
+            )
+            super().__init__(message)
+
+    class MultiPartUploadError(ObjectError):
+        """Thrown when a confirmation of an upload is rejected."""
+
+    class MultiPartUploadAlreadyExistsError(MultiPartUploadError):
+        """Thrown when trying to create a multi-part upload for an object for which another
```

Comment:
```suggestion
        """Thrown when trying to create a multipart upload for an object for which another
```
I suggest replacing "multi-part" with "multipart" everywhere for consistency.